### PR TITLE
VS License File view

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug
Fixes: https://github.com/NuGet/Home/issues/7444
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 

Add the ability to view licenses in VS. 
There are 2 ways to get to the license page right now. 
Via the Package Metadata in the Browse/Install tabs and the License AcceptanceWindow. 

In the package metadata, when a package is being browsed from an offline feed, a user will be shown a "View License" hyperlink that will open a modal dialog with the text of the license. 
Note that there is no markdown parsing as of now. 

In the License Acceptance Window, we will open expand the current window when the user clicks on it, with another text box that contains the license information.

## Testing/Validation
Tests Added: Yes, some, a lot of the work is UI.
Reason for not adding tests: A lot of the work is UI. Will add it to the vendor test cases.  
Validation done:  Yes, manual + automation.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: why use a format if you don't pass any parameters?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: This looks weird. Why make it return a `Lazy<object>`? Since it is a method and not a property is it even helpful to return a lazy which will always be `null` (because you are only returning `null`)?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml]: 💯

Comment (PatoBeltran) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml]: Make sure to test this works as expected using narrator and windows inspect object

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs]: nit: remove double `;`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs]: why 434? please make it a const with a descriptive name so it's easier to understand

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs]: same for the 400

File: src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs]: nit: remove extra space

File: src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs]: please put the magic number in a constant with a descriptive name

File: src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs]: Why use format without any parameters?

**Review by PatoBeltran — COMMENTED** [review]
Looks good 🌵 . Please attach images of the UI added for a visual review. Also, make sure to test the accessibility of the new UI by doing a pass with the narrator and making sure the UI tree looks good in windows inspect object.

**Comment by nkolev92** [comment]
licensefile-viewlicense-details
licensefile-viewsmalllicense-licenseacceptancewindow
licensefile-install-basiclicenseacceptancewindow
licensefile-viewapache-licenseacceptancewindow
licensefile-viewlicense-clickfromdetails


@PatoBeltran 
Added the screenshot. 
Any feedback on those?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: I have removed it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: The caller in this case is the UI thread, a click event. 
This method kicks off a potentially heavy operation with JTF. 
The Lazy is there to prevent someone just clicking multiple times and us doing the same expensive operation twice. 
The null is returned there because there no such thing as Lazy<void>. 

Any other ideas are appreciated/happy to discuss offline.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml]: I could use some help testing the narrator.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/LicenseAcceptanceWindow.xaml.cs]: 400 is the midwidth of both columns. 
434 is the column width + the borders + scrollviewer sizes on the left and right. 
I will extract add consts for this.

**Comment by PatoBeltran** [comment]
Couple of questions/concerns
- Is there a way to hide the license from the license acceptance window?
- What happens in the details pane when the license name is really large, are you using character trimming?
- In the license window, add some padding to the bottom, after the text has finish, so it has the same margin as the text has at the top.
- The link https://licenses.nuget.org/licenses/Apache-2.0 seems to be broken.
- The accessibility for both the license window and the expanded license acceptance window seems to be broken, I suggest doing the following:
  - Fix any issues that appear in Keros for these windows
  - Make sure you can use narrator to navigate the UI three in both windows and get everything read
  - Look into the live region event to announce to the user that the license acceptance window has expanded

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: The lazy should've been a property, not a method. 

I've changed to interlocked because that's what we use in the SolutionRestoreWorker and throughout the rest of our code.

**Review by PatoBeltran — APPROVED** [review]
I'm okay with not blocking this PR on the accessibility fixes needed as long as the PR with these work comes afterwards and is treated as P0.

**Comment by nkolev92** [comment]
>Is there a way to hide the license from the license acceptance window?

No, not currently. @karann-msft said initially it's not necesarry, he will run it by the UX board to get their opinion on it. 

> What happens in the details pane when the license name is really large, are you using character trimming?

It will wrap. I will double check that. It's not part of this change, if it's problem and not a simple fix, I will create a new issue and address it separately.

> In the license window, add some padding to the bottom, after the text has finish, so it has the same margin as the text has at the top.

Will do. 

- [x] Fixed.

> The link https://licenses.nuget.org/licenses/Apache-2.0 seems to be broken.

That's by design, as the service is not up. There are some questions about whether it will ever be, we willr eact to that. 

Regarding the accessibility concerns, I have created a new issue, https://github.com/NuGet/Home/issues/7452. 
Those will be addressed in a separate PR.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: `LoadFileAsync`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: if an api is `async` then the name should reflect the same

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs:6]: sort

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: this is not gonna make any difference, since UI thread will still be blocked for this operation to complete. Besides this back n forth will add additional unnecessary context switch. More over it also add dependency on threadpool thread which are also limited in nature.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs]: Same as mentioned before, this shouldn't be needed.

File: src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs]: why this api is declared as a task?

File: src/NuGet.Core/NuGet.Protocol/Strings.resx
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Strings.resx]: add the license  file path here

**Review by jainaashish — COMMENTED** [review]
Overall looks good, some suggestions for improvement.

Also, I'm not too comfortable with attached snapshots. Not sure if they're approved from UX. But couple of issues I see...
- Accept/Cancel buttons are not aligned on license acceptance window
- default license view window or acceptance window size seems small compare to VS size. Is it relative or fixed? Either case, should be reasonable enough since license view window contains so much texts.
- Should there be a close button on License view window?


Other than these, we also need to make sure these windows pass accessibility and works across themes. Not sure if you had already taken care of these.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: I removed the async per the other comments.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/LicenseFileText.cs]: Fixed.

File: src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs]: It's not necessary that it's a task. I made it synchronous.

File: src/NuGet.Core/NuGet.Protocol/Strings.resx
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: Added.

**Review by nkolev92 — COMMENTED** [review]
> Accept/Cancel buttons are not aligned on license acceptance window

I have not changed that. Can you clarify what you mean? They are aligned to the right side of the licenses list column.

> default license view window or acceptance window size seems small compare to VS size. Is it relative or fixed? Either case, should be reasonable enough since license view window contains so much texts.

All windows are resizable. The default sizes are the standard 450x450 across our whole product. I discussed this with @karann-msft and he suggested to not mess with until we hear feedback from customers (which so far we have not).  

> Should there be a close button on License view window?

Similar to Pato's feedback above, that was considered but was not added immediately. 
We were gonna wait for feedback, but if both of you feel strongly about it, I can it add it in a subsequent PR for this same release. 

> Other than these, we also need to make sure these windows pass accessibility and works across themes. Not sure if you had already taken care of these.

The themes are fine. The accessibility work is mostly ok, with a couple of small exceptions. Narrating works, it's the focusing that's not 100% working yet in the LicenseAcceptanceWindow. 
Been working on those with @PatoBeltran. I will do them in a separate PR as to not block this current work. Pure accessibility bugs are easy to get approval for. 

@jainaashish 

I'd like to merge this asap.  

The plan is address accessibility in a different PR. - https://github.com/NuGet/Home/issues/7452
As far as the button alignment and close license button in the license acceptance window, I can also add those is smaller side PRs.

**Pull request merged**
